source/com/vasileff/ceylon/model/Declaration.ceylon

```diff
@@ -21,7 +21,23 @@ class Declaration() of TypeDeclaration | TypedDeclaration extends Element() {
      See also https://github.com/ceylon/ceylon/issues/162"
     shared formal Integer? qualifier;
     shared formal String name;
-    shared formal Declaration? refinedDeclaration;
+
+    variable Declaration? refinedDeclarationMemo = null;
+
+    Declaration? locateRefinedDeclaration() {
+        if (! isActual) {
+            return null;
+        }
+
+        assert(is ClassOrInterface container = this.container);
+
+        return container.getMember(name)?.refinedDeclaration;
+    }
+
+    shared default
+    Declaration? refinedDeclaration
+        =>  refinedDeclarationMemo
+            else (refinedDeclarationMemo = locateRefinedDeclaration());
```

Comment:
@sadmac7000 will this ever return a value other than `null`? Maybe it would be easier if we just made `refinedDeclaration` non-optional.